* 이 노트북은 아래에 있는 MLWave에 올라온 글을 한국어로 번역하고 아래 출처의 소스코드를 더해 몇 가지 내용을 추가한 것입니다.

# [Online Learning Perceptron From MLWave](https://mlwave.com/online-learning-perceptron/)

* 원문 출처 : https://mlwave.com/online-learning-perceptron/
* 소스코드 출처
    * https://github.com/MLWave/online-learning-perceptron
    * https://github.com/MLWave/online-learning-perceptron/blob/master/perceptron.py
    * 위 소스코드를 Python3에서 실행되도록 일부 수정하였다.
    
* 온라인 학습 퍼셉트론

* 퍼셉트론 : 가장 단순한 인공신경망이다. 이 글은 1943년에 고안된 개념에서 2015년 Kaggle 경진대회(영화 리뷰가 긍정인지 부정인지 예측)로 옮겨간다. NLP 감정 분석 작업에서 하나의 인공 신경 세포가 0.95 AUC를 얻을 수 있음을 보여준다.


## 인공 신경망의 탄생 McCulloch-Pitts Neuron
          
<img src='https://s3.amazonaws.com/s3.timetoast.com/public/uploads/photos/7051972/descarga_%281%29.jpg?1477761390' width='200'>
          
          
 신경학자인 McCulloch과 논리학자인 Pitts라는 두 연구자가 1943 년의 논문 “a Logical Calculus of the Ideas Immanent in Nervous Activity”(신경 행동에서 내재적 사고의 논리적 계산) 으로 시작되었다.

인공 신경망에 기인한 네트워크를 분석하고 그들이 어떻게 간단한 논리적 기능을 하는지 보여주었다. 그들은 후에 신경 네트워크라 부르는 기술을 첫번째로 연구한 사람이다.

* 논문링크 : https://link.springer.com/article/10.1007%2FBF02478259
* [인공지능 - 위키백과, 우리 모두의 백과사전](https://ko.wikipedia.org/wiki/%EC%9D%B8%EA%B3%B5%EC%A7%80%EB%8A%A5)

McCulloch는 신경 활동에 ‘all-or-nothing’활동이 있다고 추론했다. 활성화 임계 값에 도달하거나(출력 1) 출력하지 않으면 (출력 0) 신경세포가 발화한다.
피츠 (Pitts)는 그러한 신경 학적 원칙을 사용하여 명제 논리를 포착 할 수있는 잠재력을보고 이해했다.

<img src='https://mlwave.com/wp-content/uploads/2015/01/mccullogpitts.png'>
McCulloch-Pitts 논문에서 뉴런의 회로도(looping 뉴런과 net)

* 예를 들어, 새가 씨앗과 같이 작은 물체를 먹으려고 한다. 이 행동을 표로 나타내면 다음과 같다.

|OBJECT|BROWN?|	ROUND?|	EAT?|
|:---|:---:|---:|:---|
|Seed|	1|	1|	1|
|Leaf|	1|	0|	0|
|Golf Ball|	0|	1|	0|
|Key|	0|	0|	0|


* AND 연산
McCulloch-Pitts Neuron으로 위 표를 모델링 할 수 있다. 활성화 임계 값을 1.5로 설정하면 "BROWN"및 "ROUND"속성이 모두 충족 될 때만 뉴런이 실행된다. 그러면 입력 합계는 2 (1 + 1)로 1.5의 활성화 임계 값보다 크다.


<img src='https://mlwave.com/wp-content/uploads/2015/01/pitts-neuron.png'>


## 퍼셉트론
출처 : [퍼셉트론 - 위키백과, 우리 모두의 백과사전](https://ko.wikipedia.org/wiki/%ED%8D%BC%EC%85%89%ED%8A%B8%EB%A1%A0)

<img src='http://www.enzyklopaedie-der-wirtschaftsinformatik.de/Members/wilex4/Rosen-2.jpg/image_preview'>

퍼셉트론(perceptron)은 인공신경망의 한 종류로서, 1957년에 코넬 항공 연구소(Cornell Aeronautical Lab)의 프랑크 로젠블라트 (Frank Rosenblatt)에 의해 고안되었다. 이것은 가장 간단한 형태의 피드포워드(Feedforward) 네트워크 - 선형분류기- 으로도 볼 수 있다.
퍼셉트론 알고리즘은 하드웨어로 구현 된 최초의 인공 신경망이었다.

1960년 코넬 항공 연구소의 연구원은 미 해군 연구청의 자금으로 무작위로 400개의 광전지를 하나의 퍼셉트론에 연결하여 "마크 1 퍼셉트론 (Mark 1 perceptron)"이 탄생해 이것으로 기본 이미지 인식이 가능했다.

<img src="https://upload.wikimedia.org/wikipedia/en/5/52/Mark_I_perceptron.jpeg">
이미지 출처 : https://en.wikipedia.org/wiki/Perceptron

퍼셉트론이 동작하는 방식은 각 노드의 가중치와 입력치를 곱한 것을 모두 합한 값이 활성함수에 의해 판단되는데, 그 값이 임계치(보통 0)보다 크면 뉴런이 활성화되고 결과값으로 1을 출력한다. 뉴런이 활성화되지 않으면 결과값으로 -1을 출력한다.

마빈 민스키와 시모어 페퍼트는 저서 "퍼셉트론"에서 단층 퍼셉트론은 XOR 연산이 불가능하지만, 다층 퍼셉트론으로는 XOR 연산이 가능함을 보였다.

#### 가중치

퍼셉트론은 지도학습 분류기의 일종이다. 이전 값에 대한 학습으로 예측을 한다.

퍼셉트론은 들어오는 연결에 가중치를 할당하여 작동한다. McCulloch-Pitts Neuron을 사용하여 들어오는 연결에서 값의 합계를 구하고 다음 특정 임계 값보다 높거나 낮은 지 확인했다. 내적을 구하는 대신 퍼셉트론을 사용했다. 여기에서는 들어오는 각 값에 가중치를 곱해서 합계를 구했다. 


`sum: (value1 * weight1) + (value2 * weight2)`

`weights[feature_index] += learning_rate * error * feature_value`


## 2017 수능 6월 평가원 모의고사 국어영역에 등장한 퍼셉트론
<pre>
<img src='https://upload.wikimedia.org/wikipedia/commons/thumb/b/bc/Neuron_Hand-tuned.svg/1179px-Neuron_Hand-tuned.svg.png' width=200>
* 이미지 출처 https://ko.wikipedia.org/wiki/%EC%8B%A0%EA%B2%BD_%EC%84%B8%ED%8F%AC

인간의 신경 조직을 수학적으로 모델링하여 컴퓨터가 인간처럼 기억․학습․판단할 수 있도록 구현한 것이 인공 신경망 기술이다. 신경 조직의 기본 단위는 뉴런인데, ⓐ 인공 신경망에서는 뉴런의 기능을 수학적으로 모델링한 퍼셉트론을 기본 단위로 사용한다.

<img src='https://upload.wikimedia.org/wikipedia/commons/thumb/3/31/Perceptron.svg/1052px-Perceptron.svg.png' width=300>
* 이미지 출처 https://en.wikipedia.org/wiki/Perceptron

ⓑ퍼셉트론은 입력값들을 받아들이는 여러 개의 ⓒ입력 단자와 이 값을 처리하는 부분, 처리된 값을 내보내는 한 개의 출력 단자로 구성되어 있다. 퍼셉트론은 각각의 입력 단자에 할당된 ⓓ 가중치를 입력값에 곱한 값들을 모두 합하여 가중합을 구한 후, 고정된 ⓔ 임계치보다 가중합이 작으면 0, 그렇지 않으면 1과 같은 방식으로 ⓕ 출력값을 내보낸다.

이러한 퍼셉트론은 출력값에 따라 두 가지로만 구분하여 입력값들을 판정할 수 있을 뿐이다. 이에 비해 복잡한 판정을 할 수 있는 인공 신경망은 다수의 퍼셉트론을 여러 계층으로 배열하여 한 계층에서 출력된 신호가 다음 계층에 있는 모든 퍼셉트론의 입력 단자에 입력값으로 입력되는 구조로 이루어진다.

이러한 인공 신경망에서 가장 처음에 입력값을 받아들이는 퍼셉트론들을 입력층, 가장 마지막에 있는 퍼셉트론들을 출력층이라고 한다.

<img src='https://upload.wikimedia.org/wikipedia/commons/thumb/f/f4/Honeycrisp.jpg/1280px-Honeycrisp.jpg' width=200>
* 이미지 출처 https://en.wikipedia.org/wiki/Apple


㉠ 어떤 사진 속 물체의 색깔과 형태로부터 그 물체가 사과인지 아닌지를 구별할 수 있도록 인공 신경망을 학습시키는 경우를 생각해 보자. 먼저 학습을 위한 입력값들 즉 학습 데이터를 만들어야 한다. 학습 데이터를 만들기 위해서는 사과 사진을 준비하고 사진에 나타난 특징인 색깔과 형태를 수치화 해야 한다. 이 경우 색깔과 형태라는 두 범주를 수치화하여 하나의 학습 데이터로 묶은 다음, ‘정답’에 해당하는 값과 함께 학습 데이터를 인공 신경망에 제공한다. 이때 같은 범주에 속하는 입력값은 동일한 입력 단자를 통해 들어가도록 해야 한다. 그리고 사과 사진에 대한 학습 데이터를 만들 때에 정답인 ‘사과이다’에 해당하는 값을 ‘1’로 설정하였다면 출력값 ‘0’은 ‘사과가 아니다’를 의미하게 된다.

인공 신경망의 작동은 크게 학습 단계와 판정 단계로 나뉜다. 학습 단계는 학습 데이터를 입력층의 입력 단자에 넣어 주고 출력층의 출력값을 구한 후, 이 출력값과 정답에 해당하는 값의 차이가 줄어들도록 가중치를 갱신하는 과정이다. 어떤 학습 데이터가 주어지면 이때의 출력값을 구하고 학습 데이터와 함께 제공된 정답에 해당하는 값에서 출력값을 뺀 값 즉 오차 값을 구한다. 이 오차 값의 일부가 출력층의 출력 단자에서 입력층의 입력 단자 방향으로 되돌아가면서 각 계층의 퍼셉트론별로 출력 신호를 만드는 데 관여한 모든 가중치들에 더해지는 방식으로 가중치들이 갱신된다. 이러한 과정을 다양한 학습 데이터에 대하여 반복하면 출력값들이 각각의 정답 값에 수렴하게 되고 판정 성능이 좋아진다. 오차 값이 0에 근접하게 되거나 가중치의 갱신이 더 이상 이루어지지 않게 되면 학습 단계를 마치고 판정 단계로 전환한다. 이때 판정의 오류를 줄이기 위해서는 학습 단계에서 대상들의 변별적 특징이 잘 반영되어 있는 서로 다른 학습 데이터를 사용하는 것이 좋다.
</pre>

## Online Learning Perceptron in Python

파이썬으로 표준 라이브러리를 사용해서 위의 퍼셉트론 알고리즘을 구현했기 때문에 스크립트가 PyPy에서 실행되고 3-4배의 속도향상이 있다. 여기에 사용된 알고리즘은 Kaggle 포럼에서 처음 발견 된 tinrtgu의 온라인 로지스틱 회귀 스크립트에서 큰 영감을 얻었다고 한다.

다음 경진대회에서 Vowpal Wabbit을 통한 해시트릭을 사용하였고 코드가 공개되어 있다.
* [Display Advertising Challenge - Kaggle, Beat the benchmark with less then 200MB of memory.](https://www.kaggle.com/c/criteo-display-ad-challenge/discussion/10322)
* 코드 : https://kaggle2.blob.core.windows.net/forum-message-attachments/53646/1539/fast_solution.py


## 온라인 학습
* online learning : https://en.wikipedia.org/wiki/Online_machine_learning

퍼셉트론은 온라인 학습이 가능하다(한 번에 하나씩 샘플을 통해 학습). 메모리에 전체 데이터 세트가 필요하지 않으므로(out-of-core) 더 큰 데이터 세트에 유용하다. 여기에서는 Vowpal Wabbit에서 온라인 학습 코드를 가져왔다. 

* [Vowpal Wabbit (Fast Learning)](http://hunch.net/~vw/)
* [Hashing Representations](http://hunch.net/~jl/projects/hash_reps/)

## 해싱 트릭
벡터화 해싱 트릭은 Vowpal Wabbit(John Langford)에서 시작되었다. 이 트릭은 퍼셉트론으로 들어오는 연결 수를 고정 된 크기로 설정한다. 고정 된 크기보다 낮은 숫자로 모든 원시 피처를 해싱한다. Vowpal Wabbit은 모든 데이터를 메모리로 읽어들이지 않고 모델을 훈련 시킬 수 있는 빠른 머신러닝 라이브러리다.

In [1]:
sample = "This movie sucks"
fixed_size = 1024

print(sample.split())

features = [(hash(f)%fixed_size,1) for f in sample.split()]

# list of tuples in form (feature_index,feature_value)
print(features)

['This', 'movie', 'sucks']
[(746, 1), (981, 1), (282, 1)]


### 프로그레시브 검증 손실
한 번에 하나씩 표본을 학습하면 점진적으로 train loss가 된다. 모델이 타겟을 보지않고 첫 샘플을 보고 예측을 한다.
그런 다음 예측을 대상 레이블과 비교하여 오류율을 계산 한다. 오류율이 낮으면 좋은 모델에 가깝다.

### Multiple passes
Vowpal Wabbit을 사용하면 학습률이 떨어지는 데이터 집합을 여러 번 통과시킬 수 있다.

오류율이 낮아지면 데이터 세트를 여러 번 실행할 수도 있다. 트레이닝 데이터가 선형 적으로 분리 가능한 경우, 퍼셉트론은 트레이닝 세트에서 오차가 0으로 수렴된다.

In [2]:
# opts["D"] 여기에 있는 코드에서는 고정 값으로 다음의 값을 사용한다.
2 ** 25

33554432

* 위 코드에서는 pandas, numpy, scipy같은 도구를 사용하지 않고 파이썬에 내장 된 기능만을 사용해 학습하도록 한다.
* 원래 코드는 파이썬2로 되어 있지만, 파이썬3에서 돌아가도록 코드의 일부를 수정하였다.

* 소스코드 출처 https://github.com/MLWave/online-learning-perceptron

In [3]:
import re # 정규표현식을 사용해 텍스트 데이터를 정제하기 위해
import random # 랜덤 숫자를 생성하기 위해
from math import exp, log # 지수함수와 로그함수를 사용하기 위해
from datetime import datetime # 시간을 계산하기 위해
from operator import itemgetter # 키가 아닌 값으로 max, min 값을 구할 때 사용

In [4]:
def clean(s):
    """
        Returns a cleaned, lowercased string
        텍스트 데이터를 정제하고 소문자로 변환해 준다.
    """
    return " ".join(re.findall(r'\w+', s, flags = re.UNICODE)).lower()

In [5]:
def get_data_tsv(loc_dataset,opts):
    """
    Running through data in an online manner
    Parses a tsv file for this competition 
    and yields label, identifier and features
    output:
            label: int, The label / target (set to "1" if test set)
            id: string, the sample identifier
            features: list of tuples, 
                in the form [(hashed_feature_index,feature_value)]
            
    온라인 학습 방법을 통해 데이터를 실행한다.
    tsv파일을 통해 레이블, identifier, 피처(특성)를 파싱한다.
    결과물:
        label : int, 레이블 / 대상 (테스트 집합 인 경우 "1"로 설정)
        id : 문자열, 샘플 식별자
        features : [(hashed_feature_index, feature_value)] 
                형식의 튜플 목록
    """
    for e, line in enumerate(open(loc_dataset,"rb")):
        if e > 0:
            r = line.decode('utf-8').strip().split("\t")
            id = r[0]

            if opts["clean"]:
                try:
                    r[2] = clean(r[2])
                except:
                    r[1] = clean(r[1])

            # opts["D"] = 2 ** 25 = 33554432
            # Vowpal Wabbit의 해싱트릭을 사용한다.
            # 해싱트릭은 큰 규모의 feature공간을 
            # 고정크기의 표현을 사용해 저장할 수 있게 한다.
            if len(r) == 3: # train set
                features = [(hash(f)%opts["D"],1) for f in r[2].split()]
                label = int(r[1])
            else: #test set
                features = [(hash(f)%opts["D"],1) for f in r[1].split()]
                label = 1
            
            # bigram을 사용하면 해당 피처[i]와 다음피처[i+1]를 함께 해싱한다.
            if opts["2grams"]:
                for i in range(len(features)-1):
                    features.append(
                        (hash(str(features[i][0])+str(features[i+1][0]))%opts["D"],1))
            yield label, id, features

In [6]:
def dot_product(features,weights):
    """
    Calculate dot product from features and weights
    input:
            features: A list of tuples [(feature_index,feature_value)]
            weights: the hashing trick weights filter, 
            note: length is max(feature_index)
    output:
            dotp: the dot product
    피처(특성)과 가중치로부터 내적을 구한다.
    """
    dotp = 0
    for f in features:
        dotp += weights[f[0]] * f[1]
    return dotp

In [7]:
def train_tron(loc_dataset,opts):
    start = datetime.now()
    print("\nPass\t\tErrors\t\tAverage\t\tNr. Samples\tSince Start")

    # 가중치 초기화
    if opts["random_init"]:
        random.seed(3003)
        weights = [random.random()] * opts["D"]
    else:
        weights = [0.] * opts["D"]

    # Running training passes
    # 학습 실행
    for pass_nr in range(opts["n_passes"]):
        error_counter = 0
        for e, (label, id, features) in enumerate( \
            get_data_tsv(loc_dataset,opts) ):
            
            # 퍼셉트론은 지도학습 분류기의 일종이다. 
            # 이전 값에 대한 학습으로 예측을 한다. 
            # 내적(dotproduct) 값이 임계 값보다 높거나 낮은지에 따라 
            # 초과하면 "1"을 예측하고 미만이면 "0"을 예측한다.
            dp = dot_product(features, weights) > 0.5
            
            # 다음 perceptron은 샘플의 레이블을 본다. 
            # 실제 레이블 데이터에서 위 퍼셉트론으로 구한 dp값을 빼준다.
            # 예측이 정확하다면, error 값은 "0"이며, 가중치만 남겨 둔다. 
            # 예측이 틀린 경우 error 값은 "1" 또는 "-1"이고 다음과 같이 가중치를 업데이트 한다.
            # weights[feature_index] += learning_rate * error * feature_value
            error = label - dp 
        
            # 예측이 틀린 경우 퍼셉트론은 다음과 같이 가중치를 업데이트한다.
            if error != 0:
                error_counter += 1
                # Updating the weights
                for index, value in features:
                    weights[index] += opts["learning_rate"] * error * log(1.+value)

        #Reporting stuff
        print("%s\t\t%s\t\t%s\t\t%s\t\t%s" % ( \
            pass_nr+1,
            error_counter,
            round(1 - error_counter /float(e+1),5),
            e+1,datetime.now()-start))

        #Oh heh, we have overfit :)
        if error_counter == 0 or error_counter < opts["errors_satisfied"]:
            print("%s errors found during training, halting"%error_counter)
            break
    return weights

In [9]:
def test_tron(loc_dataset,weights,opts):
    """
        output:
                preds: list, a list with
                [id,prediction,dotproduct,0-1normalized dotproduct]
    """
    start = datetime.now()
    print("\nTesting online\nErrors\t\tAverage\t\tNr. Samples\tSince Start")
    preds = []
    error_counter = 0
    for e, (label, id, features) in enumerate( \
        get_data_tsv(loc_dataset,opts) ):

        dotp = dot_product(features, weights)
        # 내적이 0.5보다 크다면 긍정으로 예측한다.
        dp = dotp > 0.5
        if dp > 0.5: # we predict positive class
            preds.append( [id, 1, dotp ] )
        else:
            preds.append( [id, 0, dotp ] )
        
        # get_data_tsv에서 테스트 데이터의 레이블을 1로 초기화 해주었음
        if label - dp != 0:
            error_counter += 1

    print("%s\t\t%s\t\t%s\t\t%s" % (
        error_counter,
        round(1 - error_counter /float(e+1),5),
        e+1,
        datetime.now()-start))

    # normalizing dotproducts between 0 and 1 
    # 내적을 구해 0과 1로 일반화 한다.
    # TODO: proper probability (bounded sigmoid?), 
    # online normalization
    max_dotp = max(preds,key=itemgetter(2))[2]
    min_dotp = min(preds,key=itemgetter(2))[2]
    for p in preds:
        # appending normalized to predictions
        # 정규화 된 값을 마지막에 추가해 준다.
        # (피처와 가중치에 대한 내적값 - 최소 내적값) / 최대 내적값 - 최소 내적값
        # 이 값이 캐글에서 0.95의 AUC를 얻을 수 있는 값이다.
        p.append((p[2]-min_dotp)/float(max_dotp-min_dotp)) 
        
    #Reporting stuff
    print("Done testing in %s"%str(datetime.now()-start))
    return preds

In [10]:
#Setting options
opts = {}
opts["D"] = 2 ** 25
opts["learning_rate"] = 0.1
opts["n_passes"] = 80 # Maximum number of passes to run before halting
opts["errors_satisfied"] = 0 # Halt when training errors < errors_satisfied
opts["random_init"] = False # set random weights, else set all 0
opts["clean"] = True # clean the text a little
opts["2grams"] = True # add 2grams

#training and saving model into weights
%time weights = train_tron("data/labeledTrainData.tsv",opts)


Pass		Errors		Average		Nr. Samples	Since Start
1		5648		0.77408		25000		0:00:19.187437
2		3161		0.87356		25000		0:00:38.114154
3		2218		0.91128		25000		0:00:57.910578
4		1643		0.93428		25000		0:01:13.906263
5		1254		0.94984		25000		0:01:29.959013
6		1038		0.95848		25000		0:01:44.738919
7		805		0.9678		25000		0:01:59.575742
8		579		0.97684		25000		0:02:14.277759
9		513		0.97948		25000		0:02:29.356437
10		464		0.98144		25000		0:02:44.361049
11		367		0.98532		25000		0:02:59.225589
12		363		0.98548		25000		0:03:19.239056
13		231		0.99076		25000		0:03:35.903281
14		203		0.99188		25000		0:03:53.428992
15		160		0.9936		25000		0:04:08.202330
16		163		0.99348		25000		0:04:27.039214
17		144		0.99424		25000		0:04:43.598832
18		168		0.99328		25000		0:04:59.653083
19		99		0.99604		25000		0:05:16.492206
20		98		0.99608		25000		0:05:31.565497
21		127		0.99492		25000		0:05:46.293131
22		81		0.99676		25000		0:06:01.460219
23		73		0.99708		25000		0:06:17.186924
24		92		0.99632		25000		0:06:33.383587
25

In [11]:
# testing and saving predictions into preds
%time preds = test_tron("data/testData.tsv",weights,opts)


Testing online
Errors		Average		Nr. Samples	Since Start
12731		0.49076		25000		0:00:17.228997
Done testing in 0:00:17.255413
CPU times: user 16.7 s, sys: 143 ms, total: 16.9 s
Wall time: 17.3 s


In [12]:
preds[:10]

[['"12311_10"', 1, 77.70179894076985, 0.6557850948318409],
 ['"8348_2"', 0, -99.60524984646412, 0.4754283296904748],
 ['"5828_4"', 0, -0.2079441541680142, 0.5765352887259396],
 ['"7186_2"', 0, -3.2577917486317594, 0.5734329831488402],
 ['"12128_7"', 1, 22.66591280431018, 0.5998025805541847],
 ['"2913_8"', 1, 68.48294143932256, 0.6464076711556087],
 ['"4396_1"', 0, -36.2515975432851, 0.5398716773602201],
 ['"395_2"', 0, -1.3169796430639447, 0.5754071776069943],
 ['"10616_1"', 0, -87.47517418666499, 0.48776704505393825],
 ['"9074_9"', 0, -23.497689420982187, 0.5528449552280901]]

In [13]:
# writing kaggle submission
# 캐글 점수 제출을 위한 서브미션 파일을 작성한다.
with open("data/submit_perceptron.csv","wb") as outfile:
    outfile.write('"id","sentiment"\n'.encode('utf-8'))
    for p in sorted(preds):
        outfile.write("{},{}\n".format(p[0],p[3]).encode('utf-8'))

In [14]:
# 캐글 스코어 0.95338
129/578

0.2231833910034602

In [15]:
# 캐글에 제출할 때 바이너리 타입으로 제출하라고 되어 있는데 여기에서는 바이너리로 변환하지 않은 채 제출을 했다.
# 그런데 캐글에서 점수가 평가되고 심지어 높은 점수로! 
# 그래서 다시 바이너리 형태의 예측값을 제출했더니 점수가 낮아졌다. 
# 바이너리로 제출한 스코어가 훨씬 낮은 0.89132 가 나왔지만 튜토리얼에 있는 방법을 사용했을 때보다 훨씬 높은 스코어다.
# 머신러닝 패키지를 사용하지 않고 파이썬의 내장 기능을 통해서 이런 점수가 나올 수 있다는 것에 놀랐다.
import pandas as pd

presult = pd.DataFrame(preds)
presult.head()

0  1          2         3
0  "12311_10"  1  77.701799  0.655785
1    "8348_2"  0 -99.605250  0.475428
2    "5828_4"  0  -0.207944  0.576535
3    "7186_2"  0  -3.257792  0.573433
4   "12128_7"  1  22.665913  0.599803

In [16]:
output_sentiment = presult[1].value_counts()
print(output_sentiment[0] - output_sentiment[1])
output_sentiment

462


0    12731
1    12269
Name: 1, dtype: int64



### 상위 벤치 마크
포럼에서 현재 가장 좋은 벤치 마크 는 Abhishek 의 로지스틱 회귀 스크립트다. 리더 보드에서 0.95 AUC 를 얻기 위해 학습과 테스트 세트의 메모리 안에서 tfidf-fit_transform을 사용한다. 여기에서는 이 변환을 건너 뛰고 온라인 해시 벡터 라이저를 사용한다.

Abhishek의 스크립트처럼 기능에서 2-gram("좋지 않은"등)을 생성한다. 2-gram을 단순히 해쉬하고 샘플 벡터에 추가한다.

또, 문자를 소문자로 바꾸고 영숫자가 아닌 것을 제거함으로써 텍스트를 조금 더 빠르게 정제한다.

### 예측
AUC를 최적화하기 위해 "1"또는 "0"예측만 제출하지 않는다. 이것은 약 0.88의 AUC를 줄 것이다. `여기에서는 0과 1 사이에서 정규화된 dotproduct를 제출한다.`

이 스크립트는 Abhishek의 솔루션 0.952 점수와 비교해 본다. 하지만 이 코드는 단지 몇 MB의 메모리를 필요로하며, tfidf 변환을 수행하지 않고 단지 2분 만에 0의 학습오류에 수렴한다다.

텍스트 정제와 2-gram이 없으면 스크립트는 60초 이내에 실행되고 0.93의 점수를 산출한다.


## 결론
"절대적인 지식은 없다. 모든 것은 확률에 의해서만 이루어진다 " - Gödel (1961)

0-1 임계 값 활성화가 적용된 매우 기본적인 퍼셉트론은 NLP 경진대회에서 잘 동작한다. 여기에서는 1957 알고리즘을 전혀 변경하지 않았다. 해싱 트릭을 없애고 비슷하거나 약간 더 좋은 점수를 받을 수 있다.

이 스크립트는 확률을 출력 할 수 없다. dotproduct에 bounded sigmoid를 사용하여 출력의 온라인 정규화 형태를 얻거나 tinrtgu의 (Vowpal Wabbit)스크립트를 더 보완해 볼 수 있을 것이다.

단일 노드 단일 레이어 퍼셉트론은 비선형 함수를 모델링 할 수 없으므로 더 나은 NLP 출력을 얻으려면 FFN(feedforward nets), recurrent nets, self-organizing maps, MLP(Multi-layer Perceptrons), word2vec 및 extreme learning machine (백프로파게이션이 없는 fast ffnets)을 봐야할 것이다.

